File: 6- Reinforcement learning/Upper Confidence Bound (UCB) /python/upper_conf_bound.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Apr  9 18:50:11 2017

UPPER CONFIDENCE BOUND

@author: Ilaria
"""

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import math

dataset = pd.read_csv('Ads_CTR_Optimisation.csv')


#implementing UCB
N = 10000
d = 10
numbers_of_selection = [0] * d
sums_of_rewards = [0] * d  
ads_selected = []    
total_reward = 0             
#all users                  
for n in range(0,N):
    ad = 0
    max_upper_bound = 0
    #all version of the ads
    for i in range(0,d):
        if(numbers_of_selection[i]> 0):
            average_reward= sums_of_rewards[i]/numbers_of_selection[i]
            delta_i = math.sqrt(3/2*math.log(n+1)/numbers_of_selection[i])
            upper_bound = average_reward + delta_i
        else:
            upper_bound = 1e400
        if upper_bound > max_upper_bound:
            max_upper_bound = upper_bound
            ad = i
    ads_selected.append(ad)
    numbers_of_selection[ad] = numbers_of_selection[ad] + 1
    reward = dataset.values[n, ad]
    sums_of_rewards[ad] = sums_of_rewards[ad] + reward
    total_reward = total_reward + reward
    
    
#visualizing the results
plt.hist(ads_selected)
plt.title('Histogram of ads selection')
plt.xlabel('ads')
plt.ylabel('Number of times each ad was selected')    
    
    

File: 6- Reinforcement learning/Upper Confidence Bound (UCB) /R/upper_conf_bound.R
#RE-INFORCEMENT LEARNING
#UPPER CONFIDENCE BOUND

#import the dataset
dataset = read.csv('Ads_CTR_Optimisation.csv')

#Implement UCB
N = 10000
d = 10
numbers_of_selections = integer(d)
sums_of_rewards = integer(d)
ads_selected = integer(0)
total_reward = 0 
#loop over all users
for(n in 1:N){
  ad = 0 
  max_upper_bound =0
  #loop over all versions of the ads
  for (i in 1:d){
    if(numbers_of_selections[i]>0){
      average_reward = sums_of_rewards[i]/numbers_of_selections[i]
      delta_i = sqrt(3/2 * log(n)/numbers_of_selections[i])
      upper_bound = average_reward + delta_i
    }
    else{
      upper_bound = 1e400
    }
    if(upper_bound > max_upper_bound){
      max_upper_bound = upper_bound
      ad = i 
    }
  }
  ads_selected = append(ads_selected,ad)
  numbers_of_selections[ad] = numbers_of_selections[ad]+1
  reward = dataset [n, ad]
  sums_of_rewards [ad] = sums_of_rewards [ad] + reward
  total_reward =  total_reward + reward
}
  

#visualizing results with an histogram
hist(ads_selected, col= 'blue', main = 'Histogram of ads selection', xlab = 'Ads', ylab = '# of time each time was selected')